raft/raft.go

```diff
@@ -565,15 +565,25 @@ func (r *raft) Step(m pb.Message) error {
 	case m.Term > r.Term:
 		lead := m.From
 		if m.Type == pb.MsgVote {
+			if r.state == StateFollower && r.checkQuorum && r.electionElapsed < r.electionTimeout {
+				r.logger.Infof("%x [logterm: %d, index: %d, vote: %x] ignored vote from %x [logterm: %d, index: %d] at term %d",
+					r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
+				return nil
+			}
 			lead = None
 		}
 		r.logger.Infof("%x [term: %d] received a %s message with higher term from %x [term: %d]",
 			r.id, r.Term, m.Type, m.From, m.Term)
 		r.becomeFollower(m.Term, lead)
 	case m.Term < r.Term:
-		// ignore
-		r.logger.Infof("%x [term: %d] ignored a %s message with lower term from %x [term: %d]",
-			r.id, r.Term, m.Type, m.From, m.Term)
+		if r.checkQuorum && (m.Type == pb.MsgHeartbeat || m.Type == pb.MsgApp) {
```

Comment:
> Seems that returning a MsgAppResp to a follower or candidate is strange. 

We don't return MsgAppResp to a follower or candidate, because only a leader would send a MsgApp or MsgHeartbeat and enter this branch. But even when sending to a leader, it's strange to send a MsgAppResp without other details like a log index.

I agree that semantically, MsgHeartbeatResp is better, but it would cause some problems for us - we intercept all heartbeat responses and send them in a single message, with no details like term numbers (although we've currently disabled our coalesced heartbeats, so it wouldn't be a problem for now). Could we add some way to distinguish heartbeat responses sent here from those sent in response to actual heartbeats? (Or maybe even a new message type - MsgHeartbeatResp causes RecentActive to be true, even though a node in this state is not completely "active".